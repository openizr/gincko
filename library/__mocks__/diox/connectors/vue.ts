/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * diox/connectors/vue mock.
 */
const connect = vi.fn(() => vi.fn(() => ({
  steps: [{ id: 'start' }, { id: 'end' }],
  loading: process.env.LOADING === 'true',
})));

export default connect;
