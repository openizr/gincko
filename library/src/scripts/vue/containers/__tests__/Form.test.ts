/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Component } from 'vue';
import { mount } from '@vue/test-utils';
import { generateRandomId } from 'sonar-ui/vue';
import Form from 'scripts/vue/containers/Form.vue';

type I18n = (label: string, values?: Record<string, string>) => string;
type ComponentApi = {
  attrs: {
    id?: string;
  };
};

jest.mock('scripts/core/Engine');
jest.mock('scripts/vue/components/Step.vue', () => ({
  render(createElement: (tag: string, api: ComponentApi) => Component): Component {
    return createElement('div', {
      attrs: {
        id: 'Step',
      },
    });
  },
  mounted(): void {
    (this as unknown as { $emit: (eventName: string) => void; }).$emit('userAction');
  },
}));
jest.mock('sonar-ui/vue', () => ({
  generateRandomId: jest.fn(() => '_abcde'),
  markdown: (value: string): string => value,
  buildClass: (...values: string[]): string => values.join(' '),
}));

describe('vue/containers/Form', () => {
  const onUserAction = jest.fn();
  const customComponents = {};

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('loading next step', async () => {
    process.env.LOADING = 'true';
    const preventDefault = jest.fn();
    Event.prototype.preventDefault = preventDefault;
    const wrapper = mount(Form, {
      propsData: {
        configuration: {
          root: 'start',
          steps: {
            start: {
              fields: [],
            },
          },
          fields: {},
        },
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    const form = wrapper.find('form');
    form.trigger('submit');
    delete process.env.LOADING;
    expect(preventDefault).toHaveBeenCalled();
    expect(wrapper.html()).toMatchSnapshot();
    // Reflects configuration change.
    wrapper.setProps({ configuration: {} });
    (wrapper.vm.$props as unknown as { i18n: I18n; }).i18n('test');
    (wrapper.vm.$props as unknown as { i18n: I18n; }).i18n('test', { test: 'test' });
    await wrapper.vm.$nextTick();
    expect(generateRandomId).toHaveBeenCalledTimes(2);
  });

  test('with active step', async () => {
    const wrapper = mount(Form, {
      propsData: {
        configuration: {
          root: 'start',
          steps: {
            start: {
              fields: [],
            },
          },
          fields: {},
        },
        activeStep: 'start',
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    await wrapper.vm.$destroy();
  });
});
