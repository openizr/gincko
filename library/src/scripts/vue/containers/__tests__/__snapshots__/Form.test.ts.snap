// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue/containers/Form loading next step 1`] = `
"<form class=\\"gincko\\">
  <div class=\\"gincko__steps\\">
    <div class=\\"ui-loader\\"></div>
  </div>
</form>"
`;

exports[`vue/containers/Form with active step 1`] = `
"<form class=\\"gincko\\">
  <div class=\\"gincko__steps\\">
    <div id=\\"start\\" index=\\"0\\" is-active=\\"true\\" custom-components=\\"[object Object]\\"></div>
    <div id=\\"end\\" index=\\"1\\" custom-components=\\"[object Object]\\"></div>
    <!---->
  </div>
</form>"
`;
